packages/wonder-stuff-core/src/error-info.js

```diff
@@ -0,0 +1,203 @@
+// @flow
+import {buildCausedByMessage} from "./build-caused-by-message.js";
+
+/**
+ * Normalized error message and stack.
+ */
+export class ErrorInfo {
+    +_message: string;
+    +_stackFrames: $ReadOnlyArray<string>;
+
+    constructor(message: string, stackFrames: $ReadOnlyArray<string>) {
+        this._message = message;
+        // Copy the stack array just to be sure our internal state won't
+        // mutate outside our control.
+        this._stackFrames = [...stackFrames];
+    }
+
+    /**
+     * The normalized error message associated with this stack trace.
+     */
+    get message(): string {
+        return this._message;
+    }
+
+    /**
+     * An array representing the stack frames of the error.
+     *
+     * Index 0 being the top of the stack.
+     */
+    get stack(): $ReadOnlyArray<string> {
+        // Copy the stack array just to be sure our internal state won't
+        // mutate outside our control.
+        return [...this._stackFrames];
+    }
+
+    /**
+     * Get the stack trace as a string.
+     */
+    toString(): string {
+        // Now construct the string value.
+        return `${this._message}\n${this._stackFrames.join("\n")}`;
+    }
+
+    /**
+     * Combine the info of an error and the error that caused it.
+     *
+     * This combines the stacks of a consequential error and the error that
+     * caused it, creating a new stack trace that incorporates the frames of
+     * both errors to tell a deeper story of what happened.
+     *
+     * For example, some work is done and error A gets thrown, this is caught
+     * and ultimately wrapped by error B, which then gets thrown. By combining
+     * the stack traces of both error A and B we can get a better idea of what
+     * happened, which can be especially useful in asynchronous code.
+     *
+     * Stack traces are combined such that common frames are deduped, and then
+     * at the point of the original throw of error A (i.e. error A's top stack
+     * frame), the deduped trace of error B is inserted to show what happened
+     * after the error was caught.
+     *
+     * So, given:
+     *
+     * cause = ErrorA, with stack,
+     *    Error: A
+     *      stackA2
+     *      stackA1
+     *      stackShared2
+     *      stackShared1
+     *
+     * and consequence = ErrorB, with stack,
+     *    Error: B
+     *       stackB2
+     *       stackB1
+     *       stackShared2
+     *       stackShared1
+     *
+     * This method would produce the equivalent of:
+     *    Error: B
+     *        caused by
+     *            Error: A
+     *      stackA2
+     *      stackA1
+     *      stackB2
+     *      stackB1
+     *      stackShared2
+     *      stackShared1
+     */
+    static fromConsequenceAndCause(
```

Comment:
**_LOVE!!_** 📖 😻

Reply:
I guess the Jane Austen version would be "Cause and Consequence" 🤣